oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/DynamicSyncContext.java

```diff
@@ -234,7 +235,8 @@ private Map<ExternalIdentityRef, SyncEntry> collectSyncEntries(@NotNull Iterable
      */
     private void collectSyncEntries(@NotNull Iterable<ExternalIdentityRef> declaredGroupRefs, long depth, @NotNull Map<ExternalIdentityRef, SyncEntry> map) throws ExternalIdentityException, RepositoryException {
         boolean shortcut = shortcut(depth);
-        for (ExternalIdentityRef ref : Iterables.filter(declaredGroupRefs, this::isSameIDP)) {
+        Iterable<ExternalIdentityRef> filtered = () -> CollectionUtils.toStream(declaredGroupRefs).filter(this::isSameIDP).iterator();
```

Comment:
How about having a util for this transformation as well?

> @NotNull
>     public static <T> Iterable<T> toIterable(@NotNull Iterable<T> iterable, @NotNull Predicate<? super T> predicate) {
>         Objects.requireNonNull(iterable);
>         Objects.requireNonNull(predicate);
>         return () -> StreamSupport.stream(iterable.spliterator(), false).filter(predicate).iterator();
>     }

Reply:
Another option could be to use `commons-collections` `IterableUtils.filteredIterable` and same goes for `Iterator`.

Reply:
Using the first proposal wouldn't work well when mutiple stream manipulations are changed.